test/extended/util/configv1shim_test.go

```diff
@@ -0,0 +1,569 @@
+			staticObjects: []runtime.Object{
+				createNetworkObject("staticObject"),
+			},
+			realObjects: []runtime.Object{
+				createNetworkObject("staticObject"),
+				createNetworkObject("realObject"),
+			},
+			watch: func(client *ConfigClientShim, ctx context.Context, opts metav1.ListOptions) (watch.Interface, error) {
+				return client.ConfigV1().Networks().Watch(ctx, opts)
+			},
+			expectedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: createNetworkObject("staticObject")},
+				{Type: watch.Added, Object: createNetworkObject("realObject")},
+			},
+		},
+		{
+			name: "merging static and real network objects, field selector match",
+			staticObjects: []runtime.Object{
+				createNetworkObject("staticObject"),
+			},
+			realObjects: []runtime.Object{
+				createNetworkObject("staticObject"),
+			},
+			watch: func(client *ConfigClientShim, ctx context.Context, opts metav1.ListOptions) (watch.Interface, error) {
+				return client.ConfigV1().Networks().Watch(ctx, opts)
+			},
+			fieldSelector: "metadata.name==staticObject",
+			expectedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: createNetworkObject("staticObject")},
+			},
+		},
+		{
+			name: "merging static and real network objects, field selector no match",
+			staticObjects: []runtime.Object{
+				createNetworkObject("staticObject"),
+			},
+			realObjects: []runtime.Object{
+				createNetworkObject("staticObject"),
+			},
+			watch: func(client *ConfigClientShim, ctx context.Context, opts metav1.ListOptions) (watch.Interface, error) {
+				return client.ConfigV1().Networks().Watch(ctx, opts)
+			},
+			fieldSelector:       "metadata.name=!staticObject",
+			expectedWatchEvents: []watch.Event{},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			configClient := fakeconfigv1client.NewSimpleClientset()
+
+			client := NewConfigClientShim(
+				configClient,
+				test.staticObjects,
+			)
+
+			// The watch request has to created first when using the fake clientset
+			resultChan, err := test.watch(client, context.TODO(), metav1.ListOptions{FieldSelector: test.fieldSelector})
+			if err != nil {
+				t.Fatalf("Expected no error, got %q instead", err)
+			}
+			defer resultChan.Stop()
+
+			// And only then the fake clientset can be populated to generate the watch event
+			for _, obj := range test.realObjects {
+				configClient.Tracker().Add(obj)
+			}
+
+			// verify the watch events
+			ticker := time.NewTicker(500 * time.Millisecond)
+			size := len(test.expectedWatchEvents)
+			eventCounter := 0
+			for i := 0; i < size; i++ {
+				select {
+				case item := <-resultChan.ResultChan():
+					diff := cmp.Diff(test.expectedWatchEvents[i], item)
+					if diff != "" {
+						t.Errorf("test '%s' failed. Results are not deep equal. mismatch (-want +got):\n%s", test.name, diff)
+					}
+					eventCounter++
+				case <-ticker.C:
+					t.Errorf("failed waiting for watch event")
+				}
+			}
+			if eventCounter != size {
```

Comment:
since we poll up to size at maximum we will not fail if we get more events than expected. This probably needs at least one extra poll from the ResultChan to detect such bug.